enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,141 @@
+superseded-by:
+
+---
+
+# On-prem: configurable DNS/LB self-hosted stack 
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+In the current implementation,  the self-hosted DNS/LB stack runs under the MCO umbrella by default in all on-prem platforms and there's no option to configure it. 
+Some customers with high production needs usually have their own external Load Balancing and DNS resolution. In addition, there are cases in which some traffic used to provide these self-hosted services is disallowed in the customer's network.                                                                                                                                                                                                                                                                                                                                  The outcome of such scenarios is that cluster's resources are spent on providing unused services and generating traffic against certain customers network design.
+
+This enhancement suggests a new method for providing the self-hosted stack.
+ 
+*Note:* as per the OCP plan to move API and Ingress to K8S SLB, this direction should be investigated deeper after MetalLb will be integrated.
+
+
+## Motivation
+
+To fully control which customers run our self hosted stack and to be able to serve the needs they really have instead of providing them in a generic way. 
+Additionally, detaching the self hosted stack configuration from the selected platform will allow new customers from UPI (platform: None) to leverage the self-hosted stack.
+
+
+### Goals
+
+- Ability to use the self-hosted stack in all on-prem platforms.
+- Ability to control the self-hosted stack configuration.
+- Retrieve the self-hosted stack status through K8S API.
+- Maintain a single code base/manifests for all the on-prem platforms.
+- Use the self-hosted stack also for UPI deployments.
+
+
+### Non-Goals
+
+## Proposal
+
+The [Keepalived and infra DNS components](https://github.com/openshift/machine-config-operator/tree/master/manifests/on-prem) on the bootstrap node will keep running through the MCO.
+However, the services on masters and workers nodes will be managed by a new operator instead of static pods.
+
+The following functionalities are expected as part of the new operator:
+* Introduce a CRD used by the operator to manage the self-hosted stack.
+* Fields in CR spec that control the self-hosted stack configuration.
+* Fields in CR status that expose self-hosted stack status (e.g: which node holds the API-VIP).
+* Run the self-hosted stack components (Keepalived, Coredns-MDNS, HAProxy, MDNS-publisher) as a K8S applications.
+
```

Comment:
I think we need more details around how the new operator will be managed - e.g I expect it will be a new second-level-operator similar to the cluster-baremetal-operator so https://github.com/openshift/enhancements/blob/master/enhancements/baremetal/an-slo-for-baremetal.md may be a good reference.

However what will be the "not in use" pattern here, e.g we want to allow use of this operator on all platforms (unlike the CBO which is disabled for all except `baremetal` - how do we detect when the operator should be running vs disabled?

Reply:
I would expect the operator to be active on any platforms that currently use on-prem. I believe some minimal set of services will always be running, even if the deployer has opted out of some of the functionality.

UPI is trickier. I almost wonder if we should defer that goal from this enhancement as it introduces some other issues (for example, this operator is still going to rely on MCO deploying some things on the node, like the DNS prepender script. Those aren't currently deployed at all on UPI). It may warrant a whole separate enhancement.